docs/v1/guidance-for-generating-p4info.md

```diff
@@ -0,0 +1,363 @@
+bit<W>` is the last element of `type_list(x)`.
+
+Note that all type names that are present as the value of a
+`type_name` field anywhere in a P4Info message must contain an entry
+describing it in the `type_info` field of the P4Info message.  Thus
+this `bitwidth` value could be considered redundant information, since
+it can be derived from the description of the type contained within
+the `type_info` field (see [this
+comment](https://github.com/p4lang/p4runtime/issues/189#issuecomment-471240515)).
+
+
+## Example 1: field with simple `bit<W>` type
+
+```
+bit<10> f1;
+
+type_list(f1) -> [bit<10>]
+
+type_name: left unset in P4Info message
+bitwidth: 10
+```
+
+Based on the P4 code snippet above, there is no need to describe any
+type inside of the `type_info` field of the P4Info message, because
+there are no named types in this code.
+
+
+## Example 2: field with user-defined `type` with no annotation
+
+```
+typedef bit<10> T1uint_t;
+@p4runtime_translation("mycompany.com/psa/v1/T1_t", 32)
+type T1uint_t T1_t;
+type T1_t T2_t;
+T2_t f2;
+```
+
+```
+Execution trace for call to type_list(f2):
+    tlist = []
+    T = declared type of object f2 in the P4 program = T2_t
+    Evaluate condition (T2_t is declared as "type B T") -> true,
+        because T2_t is declared as "type T1_t T2_t"
+    tlist = tlist + [T] -> tlist=[T2_t]
+    T = B = T1_t
+    Evaluate condition (T1_t is declared as "type B T") -> true,
+        because T1_t is declared as "type T1uint_t T1_t"
+    tlist = tlist + [T] -> tlist=[T2_t, T1_t]
+    T = B = T1uint_t
+    Evaluate condition (T1uint_t is declared as "type B T") -> false
+    Evaluate condition (T1uint_t is declared as "typedef B T") -> true,
+        because T1uint_t is declared as "typedef bit<10> T1uint_t"
+    T = B = bit<10>
+    Evaluate condition (bit<10> is declared as "type B T") -> false
+    Evaluate condition (bit<10> is declared as "typedef B T") -> false
+    tlist = tlist + [T] -> tlist=[T2_t, T1_t, bit<10>]
+    return tlist
+
+type_list(f2) -> [T2_t, T1_t, bit<10>]
+
+type_name: "T2_t"
+
+    Reason: T2_t is the first type name in type_list(f2)
+
+bitwidth: 10
+
+    Reason: Type T2_t is the first type name in type_list(f2), but it
+    has no p4runtime_translation on it, so even though T1_t does, that
+    is ignored.  Use the width 10 from bit<10>, the last element of
+    type_list(f2).
+```
+
+Based on the P4 code snippet above (copied below for easy reference),
+the value below starting with `type_info {` should be in the P4Info
+message describing the program, because of the `type` definitions.
+There is never anything put into a P4Info message because of `typedef`
+definitions in a P4 program.
+
+Note that the bit width of 10 appears in the P4Info message for any
+`type`s "built on top of" a `bit<10>`, _unless that type has its own
+`p4runtime_translation` annotation_.
+
+```
+typedef bit<10> T1uint_t;
+@p4runtime_translation("mycompany.com/psa/v1/T1_t", 32)
```

Comment:
Nit-picking: I would expect the URI to be `"p4.org/psa/v1/<type_name>"` for types defined in psa.p4 (`"company.com/<proprietary_arch>/<type_name>"` for types defined in a proprietary architecture). For types defined in a P4 program, I would expect something like `"company2.com/[<arch>/]<P4_program>/<type_name>"`, with company2 being the device user and company being the vendor. I know I put in the spec that the arch name would ideally always be part of the URI, but what I had in mind were types introduced by the architecture specifically. For types defined by a user in their program, the architecture name is not as important as something identifying the particular P4 program the type is defined in. If a company defines a type in a separate p4 file (that they include in all their P4 programs), they could go with something like `"company2.com/std/<type_name>"`.

Reply:
Nit-pick away!  I have added a section near the beginning of this document attempting to capture your recommendations here, and stating them as recommendations.  Feel free to review and comment on it after I have committed it.

Reply:
I have replaced all example URIs of hte form "mycompany.com/psa/v1/<type_name>" with "mycompany.com/myco_p4lib/v1/<type_name>".

Reply:
I meant "Nit-pick away!" in the sense of "go ahead and do it!", not "go away".